numba/core/lowering.py

```diff
@@ -926,6 +938,65 @@ def _lower_call_RecursiveCall(self, fnty, expr, signature):
             )
         return res
 
+    def _lower_call_FunctionType(self, fnty, expr, signature):
+        self.debug_print("# calling first-class function type")
+        sig = types.unliteral(signature)
+        if not fnty.check_signature(signature):
+            # value dependent polymorphism?
+            raise UnsupportedError(
+                f'mismatch of function types:'
+                f' expected {fnty} but got {types.FunctionType(sig)}')
+        ftype = fnty.ftype
+        argvals = self.fold_call_args(
+            fnty, sig, expr.args, expr.vararg, expr.kws,
+        )
+        func_ptr = self.__get_function_pointer(ftype, expr.func.name, sig=sig)
+        res = self.builder.call(func_ptr, argvals, cconv=fnty.cconv)
+        return res
+
+    def __get_function_pointer(self, ftype, fname, sig=None):
+        from numba.core.function_type import lower_get_wrapper_address
+
+        llty = self.context.get_value_type(ftype)
+        fstruct = self.loadvar(fname)
+        addr = self.builder.extract_value(fstruct, 0,
+                                          name='addr_of_%s' % (fname))
+
+        fptr = cgutils.alloca_once(self.builder, llty,
+                                   name="fptr_of_%s" % (fname))
+        with self.builder.if_else(
+                cgutils.is_null(self.builder, addr),
+                likely=False) as (then, orelse):
+            with then:
+                self.init_pyapi()
```

Comment:
Is this branch being exercised by any test? I tried to cause this branch to error out (i.e. storing NULL as the function pointer) but none of the test would fail.

Reply:
No, addr should be never be zero here.
This branch is exercised in the case of some bug and it is implemented in order to fail gracefully with a RuntimeError exception rather than a segfault.

Reply:
I meant this branch in the `with then`. Precisely the segment https://github.com/Quansight/numba/blob/6d361f7a78c86725e7307ba83269de26a01fae58/numba/core/lowering.py#L971-L995. I cannot find a test that is executing the compiled code. So I tried to set the address `addr2` (at https://github.com/Quansight/numba/blob/6d361f7a78c86725e7307ba83269de26a01fae58/numba/core/lowering.py#L993) to zero to see if a test would fail, but no test failed. I think a new test is needed to exercise this segment.

Reply:
Added a new test that exercises the above-mentioned branch.
See test_zero_address in test_function_type, in particular, the Bad2Good example.